pkg/envoy/registry/services.go

```diff
@@ -27,34 +34,162 @@ func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service
+				svc := event.NewObj.(*v1.Service)
+				k.handleServiceUpdate(svc)
+			case announcements.ServiceDeleted:
+				svc := event.OldObj.(*v1.Service)
+				k.handleServiceDelete(svc)
+			}
+			events.GetPubSubInstance().Publish(events.PubSubMessage{
+				AnnouncementType: announcements.ScheduleProxyBroadcast,
+			})
+		}
 	}
+}
 
-	services, err := listServicesForPod(pod, k.KubeController)
+func (k *KubeProxyServiceMapper) handlePodUpdate(pod *v1.Pod) {
+	if pod == nil {
+		return
+	}
+	cn, err := getCertCommonNameForPod(*pod)
 	if err != nil {
-		return nil, err
+		log.Error().Err(err).Msgf("ignoring updated pod %s/%s", pod.Namespace, pod.Name)
+		return
 	}
 
-	if len(services) == 0 {
-		return nil, nil
-	}
+	services := listServicesForPod(pod, k.kubeController)
 
 	meshServices := kubernetesServicesToMeshServices(services)
 
 	servicesForPod := strings.Join(listServiceNames(meshServices), ",")
 	log.Trace().Msgf("Services associated with Pod with UID=%s Name=%s/%s: %+v",
 		pod.ObjectMeta.UID, pod.Namespace, pod.Name, servicesForPod)
 
-	return meshServices, nil
+	k.cacheLock.Lock()
+	k.servicesForCN[cn] = meshServices
+	k.cacheLock.Unlock()
+}
+
+func (k *KubeProxyServiceMapper) handlePodDelete(pod *v1.Pod) {
+	if pod == nil {
+		return
+	}
+	cn, err := getCertCommonNameForPod(*pod)
+	if err != nil {
+		log.Error().Err(err).Msgf("ignoring deleted pod %s/%s", pod.Namespace, pod.Name)
+		return
+	}
+	k.cacheLock.Lock()
+	delete(k.servicesForCN, cn)
+	k.cacheLock.Unlock()
+}
+
+func (k *KubeProxyServiceMapper) handleServiceUpdate(svc *v1.Service) {
+	if svc == nil {
+		return
+	}
+	updatedSvc := service.MeshService{
+		Name:      svc.Name,
+		Namespace: svc.Namespace,
+	}
+
+	pods := listPodsForService(svc, k.kubeController)
+	for _, pod := range pods {
+		cn, err := getCertCommonNameForPod(pod)
+		if err != nil {
+			log.Error().Err(err)
+			continue
+		}
+		alreadyCached := false
+		k.cacheLock.RLock()
+		for _, cachedSvc := range k.servicesForCN[cn] {
+			if cachedSvc == updatedSvc {
+				alreadyCached = true
+				break
+			}
+		}
+		k.cacheLock.RUnlock()
+		if !alreadyCached {
+			k.cacheLock.Lock()
+			k.servicesForCN[cn] = append(k.servicesForCN[cn], updatedSvc)
+			k.cacheLock.Unlock()
+		}
+	}
+}
+
+func (k *KubeProxyServiceMapper) handleServiceDelete(svc *v1.Service) {
+	if svc == nil {
+		return
+	}
+	deleted := service.MeshService{
+		Name:      svc.Name,
+		Namespace: svc.Namespace,
+	}
+	k.cacheLock.RLock()
+	for cn, svcs := range k.servicesForCN {
+		var rem []service.MeshService
+		for _, s := range svcs {
+			if s == deleted {
+				continue
+			}
+			rem = append(rem, s)
+		}
+		k.cacheLock.RUnlock()
+		k.cacheLock.Lock()
+		k.servicesForCN[cn] = rem
+		k.cacheLock.Unlock()
+		k.cacheLock.RLock()
+	}
+	k.cacheLock.RUnlock()
+}
+
+// ListProxyServices maps an Envoy instance to a number of Kubernetes services.
+func (k *KubeProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
+	k.cacheLock.RLock()
```

Comment:
This implementation is basically the same as a kubernetes cache, we just switch the key from service name to certCommonName. 
For you to keep in mind that misuse of these apis will run with the same locking contention issues that we currently have using K8s caches (it's just an fyi, the fact that things are out of catalog makes every package have to deal with its own locking now).